distsql: remove stored context from FlowCtx and processors

asubiotto opened the pull request:
Resolves #12346.

This change results in passing down a context from the server entry
points to the processors instead of storing contexts on structs at
initialization and accessing them later.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13514)
<!-- Reviewable:end -->

Commit:
distsql: remove stored context from FlowCtx and processors

Resolves #12346.

This change results in passing down a context from the server entry
points to the processors instead of storing contexts on structs at
initialization and accessing them later.

tamird commented:
Not LGTM'ing because @raduberinde's the authority, but looks good.

---

Reviewed 28 of 28 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow_scheduler.go, line 47 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9eRc-LWEr7D-oXmu:-KcZ9eRc-LWEr7D-oXmv:bj6bgz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/flow_scheduler.go#L47)):*
> ```Go
> }
> 
> // runFlowNode stores a flow to run and a context to run it with.
> ```

leave a TODO to see if this can be rearranged to avoid the need for this

---

*[pkg/sql/distsqlrun/joinerbase.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9lqu_5GKO6aMxdIP:-KcZ9lqu_5GKO6aMxdIQ:bdjb7h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/joinerbase.go#L121)):*
> ```Go
> }
> 
> func (jb *joinerBase) ctxWithJoinerTag(ctx context.Context) context.Context {
> ```

why does this need to be a method?

---

*[pkg/sql/distsqlrun/server.go, line 156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZA1Ku1OYb11ROLS-Z:-KcZA1Kv7C-VaPtmX3HX:b-pgmd8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L156)):*
> ```Go
> 		return nil, err
> 	}
> ```

random newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-:-KcZAEpZedTT_K4Rw3Qj:b-ikgzr0)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinerbase.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9lqu_5GKO6aMxdIP:-KcZBJQBwLchvxQYg2W7:b1yhaj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/joinerbase.go#L121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why does this need to be a method?
</blockquote></details>

So that any types which embed `joinerBase` can just call this to keep the log tag consistent across joiners (currently `hashJoiner` and `mergeJoiner`). Not sure whether it's a good idea but it avoids programmer error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinerbase.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9lqu_5GKO6aMxdIP:-KcZHie9LvhHlZQ6jCRS:bt4ccyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/joinerbase.go#L121)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

So that any types which embed `joinerBase` can just call this to keep the log tag consistent across joiners (currently `hashJoiner` and `mergeJoiner`). Not sure whether it's a good idea but it avoids programmer error.
</blockquote></details>

hm, OK. nit: s/jb //

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for working on this Alfonso! Looks good, just a couple of comments.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow_scheduler.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZL9LZfMFWMchEby4o:-KcZL9L_sBle_sMSwf3m:b-iloz29) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/flow_scheduler.go#L48)):*
> ```Go
> 
> // runFlowNode stores a flow to run and a context to run it with.
> type runFlowNode struct {
> ```

Why "Node"? Maybe flowWithCtx

---

*[pkg/sql/distsqlrun/joinerbase.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9lqu_5GKO6aMxdIP:-KcZLPlpZQKHd3SC28k1:b-8ag2qs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/joinerbase.go#L121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, OK. nit: s/jb //
</blockquote></details>

Honestly I think using the same name was more of an omission, it's better if we distinguish "MergeJoiner" and "HashJoiner" (or we can leave "Joiner" for the hash joiner).

---

*[pkg/sql/distsqlrun/server.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZJepXG-B703vXyVyj:-KcZJepXG-B703vXyVyk:bgja111) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L72)):*
> ```Go
> }
> 
> func (ds *ServerImpl) setupFlow(
> ```

document that the returned context contains a span that needs to be finished (via flow.Cleanup)

---

*[pkg/sql/distsqlrun/server.go, line 116 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZJpL19nPjmJOUvCvT:-KcZJpL19nPjmJOUvCvU:b-flk6q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L116)):*
> ```Go
> // output stream to the given RowReceiver. The flow is not started. The flow
> // will be associated with the given context.
> func (ds *ServerImpl) SetupSyncFlow(
> ```

same, mention the context and the span

---

*[pkg/sql/distsqlrun/server.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZKBoRzjETUjPe99sD:-KcZKBoRzjETUjPe99sE:bakxhn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L186)):*
> ```Go
> 	log.VEventf(ctx, 1, "connected inbound stream %s/%d", flowID.Short(), streamID)
> 	defer ds.flowRegistry.FinishInboundStream(streamInfo)
> 	return ProcessInboundStream(ctx, stream, msg, streamInfo.receiver)
> ```

FlowCtx had more annotations.. I guess just the flow id would be missing; we can add the tag here, but perhaps it would be better to set up an `AmbientContext` in `flow` (or `FlowCtx`) and use that to annotate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-:-KcZJXxos0DQE8OjQQ-v:bcjxdw7)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 21 of 28 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow_scheduler.go, line 47 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9eRc-LWEr7D-oXmu:-KccvqCAMGHkIwmXaoZO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/flow_scheduler.go#L47)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

leave a TODO to see if this can be rearranged to avoid the need for this
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/flow_scheduler.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZL9LZfMFWMchEby4o:-KccvprRmokjUp8oilI-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/flow_scheduler.go#L48)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why "Node"? Maybe flowWithCtx
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinerbase.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZ9lqu_5GKO6aMxdIP:-Kccw7zDwOK0hjKfOPa0:b-byyvde) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/joinerbase.go#L121)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Honestly I think using the same name was more of an omission, it's better if we distinguish "MergeJoiner" and "HashJoiner" (or we can leave "Joiner" for the hash joiner).
</blockquote></details>

Removed and added different log tags for the joiners.

---

*[pkg/sql/distsqlrun/server.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZJepXG-B703vXyVyj:-KccyegIBl_GZTDVuuCF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L72)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

document that the returned context contains a span that needs to be finished (via flow.Cleanup)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/server.go, line 116 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZJpL19nPjmJOUvCvT:-Kccyf755oPJO7RPT0cs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L116)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

same, mention the context and the span
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/server.go, line 156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZA1Ku1OYb11ROLS-Z:-Kccyl_4jBwH9NMqVYHR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L156)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

random newline
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/server.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-KcZKBoRzjETUjPe99sD:-KccziMFl3WkXRA2pviw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2069e359ea53ac0eeb7c37488cb4a195ee318e3/pkg/sql/distsqlrun/server.go#L186)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

FlowCtx had more annotations.. I guess just the flow id would be missing; we can add the tag here, but perhaps it would be better to set up an `AmbientContext` in `flow` (or `FlowCtx`) and use that to annotate?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great!

---

Review status: 21 of 28 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13514#-:-KcdM5GZrC_df4-UX-Oa:bioon7n)*
<!-- Sent from Reviewable.io -->
